specification.rst

```diff
@@ -0,0 +1,229 @@
+=======================
+Nix Flake Specification
+=======================
+----------------------
+Version 0.0.0
+----------------------
+
+..  DANGER::
+    This specification covers a prototype implementation of the
+    specified interface. It has not yet been adopted by the community
+    and is more subject to change and/or abandonment than usual.
+
+Introduction
+==============
+
+.. include:: description.rst
+
+This specification formalizes the technical interface for `Nix`
+flakes. It includes of formats and semantics for flake.nix_,
+describing the metadata and members of the flakes, and for
+flake.lock, a "lock file" pinning the flake's dependencies for a
+particular evaluation. It also defines the format and semantics of the
+files defining the individual flake members.
+
+flake.nix
+===========
+The `flake.nix` file describes the flake being defined, including
+metadata and enumerating its members. Here is an example illustrating
+the format:
+
+.. code-block:: nix
+
+   {
+     metadata = {
+       name = "flake-example";
+
+       version = { major = 1; patch = 1; }; # i.e. semver 1.0.1
+
+       description = "A fake flake to illustrate flake.nix";
+     };
+
+     dependencies = {
+       foo = {
+         good-versions = [ { major = 2; minor = 1; } ];
+
+         bad-version-ranges = [ { upper-bound = {
+           version.major = 2;
+
+           included = false;
+         } } ];
+
+         # So, we depend on "foo", we know that versions prior to
+         # 2.0.0 won't work, and we expect 2.1.x and higher (but
+         # lower than 3.x) will work, but we're not sure about 2.0
+         # or anything 3.x or higher.
+       };
```

Comment:
If each flake has their own rule set like this then it's going to make the dependency solver's life quite difficult. I think that this has been shown by the cabal approach.

An alternative approach would be to only list a known good version in the flake.nix and flakes are *never* allowed to break backward-compatibility. It backward-compat is broken it's a bug and can be recorded as a bad version in a central registry (or have the library mark the version as broken somehow).

But you have to break back-compat some times you would argue; in that case, either a v2/ version and create a proxy from v1 to v2. Or create a new package, this will have the advantage of clearly showing that it's a new thing.

All of this is not my idea. See the [go modules](https://github.com/golang/go/wiki/Modules) for more details on the approach.

Another nice property of that approach is that .lock files are not necessary as the dependency solver is reproducible even if new versions are released.

Reply:
If you don't have time to watch the spec-ulation video maybe just read this core takeaway http://blog.ezyang.com/2016/12/thoughts-about-spec-ulation-rich-hickey/ . 

So it would seem one would need to incorporate the concept of a namespace into flakes that exposes a set of [named] derivations. If the author/user don't like that set of [named] derivations in a namespace later down the line one creates a new namespace with an optimal set of [named] derivations. Yet never removing the original names. 

We're faced with this problem in Fractalide too. Our solution (at least within the rust side of things, later to be brought into the racket side) is to only have one name per agent - no versioning whatsoever. We'd like to introduce a tool that checks the data types on the agents input and output ports, and if any of that changes after an agents is declared `stable` it doesn't pass the test and isn't merged.

(So a slight FYI tangent, a fractalide agent has a 5 phase life cycle `raw`, `draft`, `stable`, `deprecated`, `legacy`. An agent is in a `raw` state if only the author is the user. As soon as some bleeding edge 3rd party decides that agent is usable then it becomes `draft`. The user knows that he's using unstable software and is prepared to change the upstream and downstream agents to suit the mutated agent. It is expected that the agent could undergo significant mutation. Should the agent gain a number of users and the author decides it is solid enough, it enter `stable`. Any modification to the agent's public API is hostile behaviour and is forbidden. The agent's internals can obviously be optimised. Should users generally stop using the agent in favour of another then the agent becomes `deprecated`, a warning is printed to the screen with a suitable replacement candidate. A `legacy` agent *might* have the implementation removed and replaced by an assertion.  At no point after `stable` may names be reused or changed.)

_Ideally this tooling can be used across nixpkgs and fractalide_
The tooling would check for a number of things:
* ensure a phase is not reverted
* check that agent/derivation names do not change after `stable`
* check that ports and data types remain the same
* check that there is no agent removal after `stable` phase
* etc

This tool could be run at PR time, or incorporated into hydra, something to that effect.

In Fractalide we've incorporated a namespace (what we call a `fractal`). This will give us the flexibility to implement the points listed in blog post above.

The key point here is to allow for flexibility in the early stages of code (gain feedback from early users) and ensure things don't become user hostile later down the line, yet retain the flexibility of accretion by exposing new namespaces that consumers can migrate to *in their own time*.

Ideally we hope for this flake concept to be flexible enough to be used in the above use case. If flakes don't go in this direction we'll need to roll our own eventually.

Reply:
OK, I've watched the talk and reviewed the go documentation, and thoughts are swirling right now. I'll hopefully have an update later this week about where I think we should go during this initial phase... For now I do want to reiterate that this is just an initial phase, and until/unless we reach RFC we should not expect any of these details to be decided for the community/functionality as a whole.